libcst/metadata/scope_provider.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+import abc
+import builtins
+from collections import defaultdict
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Dict, Iterator, List, MutableMapping, Optional, Tuple, Type, Union
+
+import libcst as cst
+from libcst.metadata.base_provider import BatchableMetadataProvider
+from libcst.metadata.expression_context_provider import (
+    ExpressionContext,
+    ExpressionContextProvider,
+)
+
+
+@dataclass(frozen=True)
+class Access:
+    node: cst.Name
+    scope: "Scope"
+
+
+class BaseAssignment:
+    name: str
+    scope: "Scope"
+    __accesses: List[Access]
+
+    def __init__(self, name: str, scope: "Scope") -> None:
+        self.name = name
+        self.scope = scope
+        self.__accesses = []
+
+    def record_access(self, access: Access) -> None:
+        self.__accesses.append(access)
+
+    @property
+    def accesses(self) -> Tuple[Access, ...]:
+        # we don't want to publicly expose the mutable version of this
+        return tuple(self.__accesses)
+
+
+class Assignment(BaseAssignment):
+    node: cst.CSTNode
+
+    def __init__(self, name: str, scope: "Scope", node: cst.CSTNode) -> None:
+        self.node = node
+        super().__init__(name, scope)
+
+
+class BuiltinAssignmemt(BaseAssignment):
+    node: None = None
+
+
+class Scope(abc.ABC):
+    parent: "Scope"
+    globals: "GlobalScope"
+    _assignments: MutableMapping[str, List[BaseAssignment]]
+
+    def __init__(self, parent: "Scope") -> None:
+        super().__init__()
+        self.parent = parent
+        self.globals = parent.globals
+        self._assignments = defaultdict(list)
+
+    def record_assignment(self, name: str, node: cst.CSTNode) -> None:
+        self._assignments[name].append(Assignment(name=name, scope=self, node=node))
+
+    def _contains_in_parent_scope(self, name: str) -> bool:
+        return name in self.parent
+
+    def _get_from_parent(self, name: str) -> Tuple[BaseAssignment, ...]:
```

Comment:
This was called `_get_as_parent` in the AST version because it fetches `name` when you're the parent scope.

(Opinion) While `_get_as_parent` isn't a great name, `_get_from_parent` is even more confusing to me, because it doesn't (usually) get the item from the parent.

Reply:
What about `_getitem_from_self_or_parent`?